dep(api): update docker binary version to 19.03.12

olljanat opened the pull request:
Update Docker binary version to 19.03.x for both Linux and Windows which adds support for Docker Stack version 3.8 https://docs.docker.com/compose/compose-file/

Because Docker, Inc does not any more publish Windows binaries separately I change script to use binaries published by Microsoft. Links to those can be found from https://go.microsoft.com/fwlink/?LinkID=825636&clcid=0x409

Closes #3206 #3036

olljanat commented:
@deviantony PTAL

deviantony commented:
@olljanat last time we tried to bump the version this was causing issues with TLS enabled endpoints.

See https://github.com/portainer/portainer/issues/3206#issuecomment-537704606

I'll add this to our backlog for investigation though.

olljanat commented:
@deviantony it looks to be that on 18.09 `--tlscacert` defaults to empty but on 19.03 it defaults to value which is visible on `docker --help`

It can be tested very easily even without TLS configuration.

On 18.09:
```
$ docker --tls ps
Cannot connect to the Docker daemon at tcp://localhost:2376. Is the docker daemon running?
```

On 19.03:
```
$ docker --tls ps
unable to resolve docker endpoint: open /home/user/.docker/ca.pem: no such file or directory
```

Solution is force `--tlscacert` to empty when `--tlsverify` is not used so it works same way on both versions:
```
$ docker --tlscacert "''" --tls ps
Cannot connect to the Docker daemon at tcp://localhost:2376. Is the docker daemon running?
```

I updated that to this PR now and also updated Docker CLI for Windows.

Note! I have not tested to build this one as I was planned to use CI for it but looks that it does not work.

/cc: @itsconquest

deviantony commented:
@olljanat perfect ! Thanks for the update. I'll trigger a CI run and we'll be able to test this one.

deviantony commented:
/azp run

deviantony reviewed (approved):
LGTM

I've tested the deployment of a stack using the stack file below against a Docker API enabled endpoint:

```yaml
version: "3.8"
services:
  worker:
    image: nginx
    deploy:
      mode: replicated
      replicas: 6
      placement:
        max_replicas_per_node: 1
```

We'll now need to ensure the following:

- [ ] Changes are reflected on the agent for Edge stack support via https://github.com/portainer/agent/issues/158
- [ ] Test this PR against agent and Edge agent endpoints
- [ ] Test this PR on the Windows platform

deviantony reviewed (approved):
Went for another round of review after updating the agent via https://github.com/portainer/agent/pull/159 and it looks like the Compose v3.8 is supported through stacks deployment (only tested on Edge agent and Docker API, will need to be tested against an agent endpoint) as well as Edge stack deployment.

LGTM

@olljanat I know we're late on that one but do you mind resolving the conflict? If you can't / do not have the time for it, we can take care over it. Sorry again for the delay.

olljanat commented:
@deviantony rebased and fixed conflict

deviantony commented:
@olljanat awesome, thank you!

itsconquest commented:
@olljanat i've just done a review and it LGTM, great work!! Are you able to resolve the merge conflict for me? Then I can merge it for you

olljanat commented:
@itsconquest  done

deviantony commented:
Thanks again for that one @olljanat ! It tooks us some time but it made it to the codebase :-)
